File: 0090_subsetII/main.cpp
#include <vector>
#include <algorithm>

class Solution {
private:
    void backtrack(std::vector<int>& nums, std::vector<int>& subset, std::vector<std::vector<int>>& res, int startIdx) {
        if(startIdx == nums.size()) {
            res.push_back(subset);
            return;
        }

        subset.push_back(nums[startIdx]);
        backtrack(nums, subset, res, startIdx + 1);
        subset.pop_back();

        while(startIdx + 1 < nums.size() && nums[startIdx] == nums[startIdx + 1])
            ++startIdx;
        backtrack(nums, subset, res, startIdx + 1);

    }
public:
    std::vector<std::vector<int>> subsetsWithDup(std::vector<int>& nums) {
        std::vector<std::vector<int>> res;
        std::vector<int> subset;
        std::sort(nums.begin(), nums.end());
        backtrack(nums, subset, res, 0);
        return res;
    }
};

File: 0090_subsetII/main.py
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        def backtrack(startIdx, subset):
            if startIdx == len(nums):
                res.append(subset[:])
                return
            
            # All subsets that include nums[i]
            subset.append(nums[startIdx])
            backtrack(startIdx + 1, subset)
            subset.pop()

            # All subsets that don't include nums[i]
            while startIdx + 1 < len(nums) and nums[startIdx] == nums[startIdx + 1]:
                startIdx += 1
            backtrack(startIdx + 1, subset)

        backtrack(0, [])
        return res
    
class Solution1:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        def backtrack(startIdx, subset):
            res.append(subset[:])
            for i in range(startIdx, len(nums)):
                if i > startIdx and nums[i] == nums[i-1]:
                    continue
                subset.append(nums[i])
                backtrack(i + 1, subset)
                subset.pop()
        
        backtrack(0, [])
        return res
